### PR TITLE
AS3 override feature in CIS

### PR DESCRIPTION
**Pull request by snatra**
Problem:
Need a feature that overrides existing AS3 configuration with a new user-defined AS3 configuration, using k8s ConfigMap. So the administrator will be able to modify the existing virtual server configuration incrementally as needed without having to overwrite/delete the existing one

Solution:
Implemented AS3 override feature in CIS, this feature enables administrators to create a virtual server in a controlled and managed partition manually and add metadata so that f5-cccl will not create a new virtual server.

affects-branches: master

docker-image: snatra27/k8s-bigip-ctlr:cis-override-json-2

### CODE REVIEW

**Comment by ssurenr** [comment]
Removed 1.11-stable in affected branches as this is 1.12 feature.

File: cmd/k8s-bigip-ctlr/main.go [hunk 1]
```diff
@@ -188,7 +189,8 @@ func _init() {
 	// TODO: Rephrase agent functionality
 	agent = bigIPFlags.String("agent", "cccl",
 		"Optional, when set to as3, orchestration agent will be AS3 instead of CCCL")
-
+	overrideAS3Decl = bigIPFlags.String("override-as3-declaration", "",
+		"Optional, when set, controller will override existing as3 declaration with configured one here")
```
[the remark this answers is not in the document]
Comment (snatra) [reply cmd/k8s-bigip-ctlr/main.go]: Done

File: cmd/k8s-bigip-ctlr/main.go [hunk 2]
```diff
@@ -191,7 +192,9 @@ func _init() {
 	// TODO: Rephrase agent functionality
 	agent = bigIPFlags.String("agent", "cccl",
 		"Optional, when set to as3, orchestration agent will be AS3 instead of CCCL")
-
+	overrideAS3Decl = bigIPFlags.String("override-as3-declaration", "",
+		"Optional, when set, the JSON key/values from this ConfigMap will override"+
+			" key/values from internally generated AS3 declaration")
```
Comment (trinaths) [inline cmd/k8s-bigip-ctlr/main.go]: We can improve this helper statement to be more informational. 

"Optional, provide Namespace where the ConfigMap is existing and name of that ConfigMap in format `<namespace>/<configmap-name>`. When set the JSON from ConfigMap will override internally generated AS3 declaration."

+ documentation update required.

File: pkg/appmanager/as3Manager.go [hunk 3]
```diff
@@ -587,15 +587,14 @@ func (appMgr *Manager) getCertFromConfigMap(cfgmap string) {
...
+		log.Debugf("[as3_log] Invalid trusted-certs-cfgmap option provided.")
 	} else {
 		certs := ""
-		namespace := namespaceCfgmapSlice[0]
-		cfgmapName := namespaceCfgmapSlice[1]
-		cm, err := appMgr.kubeClient.CoreV1().ConfigMaps(namespace).Get(cfgmapName, metaV1.GetOptions{})
+		cm, err := appMgr.getConfigMapUsingNamespaceAndName(namespaceCfgmapSlice[0], namespaceCfgmapSlice[1])
 		if err != nil {
-			log.Debugf("[AS3] Reading certificate from configmap error: %v", err)
+			log.Errorf("[AS3] ConfigMap not found for this name: %v and namespace: %v, error: %v",
```
Comment (trinaths) [inline pkg/appmanager/as3Manager.go]: can be, 
ConfigMap '%v' not found in namespace '%v'. Error: %v.

File: pkg/appmanager/as3Manager.go [hunk 4]
```diff
@@ -833,6 +840,37 @@ func (appMgr *Manager) getUnifiedAS3Declaration(as3CfgmapDecl as3Declaration, ro
...
 	}
+
+	// Override AS3 Config if AS3 Override is enabled
+	if appMgr.OverrideAS3Decl != "" {
+		// Fetch configMap's namespace and name
+		overrideAS3Decl := strings.Split(appMgr.OverrideAS3Decl, "/")
+		// Fetch configMap using namespace and name as an input
+		cfgMap, err := appMgr.getConfigMapUsingNamespaceAndName(overrideAS3Decl[0], overrideAS3Decl[1])
+		if err != nil {
+			log.Errorf("[AS3] ConfigMap not found for this name: %v and namespace: %v, error: %v",
```
Comment (trinaths) [inline pkg/appmanager/as3Manager.go]: ConfigMap '%v' not found in namespace '%v'. Error: %v.

File: cmd/k8s-bigip-ctlr/main.go [hunk 5]
```diff
@@ -191,7 +192,9 @@ func _init() {
 	// TODO: Rephrase agent functionality
 	agent = bigIPFlags.String("agent", "cccl",
 		"Optional, when set to as3, orchestration agent will be AS3 instead of CCCL")
-
+	overrideAS3Decl = bigIPFlags.String("override-as3-declaration", "",
+		"Optional, when set, the JSON key/values from this ConfigMap will override"+
```
Comment (trinaths) [inline cmd/k8s-bigip-ctlr/main.go]:
```suggestion
		Optional, provide Namespace where the ConfigMap is existing and name of that ConfigMap in format `<namespace>/<configmap-name>`. when set, the JSON key/values from this ConfigMap will override key/values from internally generated AS3 declaration.+
```

File: pkg/appmanager/as3Manager.go (hunk 4 above)
Comment (snatra) [reply pkg/appmanager/as3Manager.go]: Done

File: cmd/k8s-bigip-ctlr/main.go (hunk 5 above)
Comment (snatra) [reply cmd/k8s-bigip-ctlr/main.go]: Done

File: pkg/appmanager/as3Manager.go (hunk 3 above)
Comment (snatra) [reply pkg/appmanager/as3Manager.go]: Done

**Pull request merged**